README
RTEMS GeSys ("Generic System") 
==============================

RTEMS GeSys is a general purpose application.
It is configured to support most of RTEMS' features
(all managers, networking, APIs) with generous limits.

The GeSys Makefile contains special "magic" to enforce
GeSys to be linked with most parts of the core system
libraries (libc, libstdc++, libgcc, librtemsbsp,
librtemscpu etc.) including full C++ support.
Additional libraries can be included by listing them
in the Makefile (add to OPT_LIBRARIES). Individual
objects may be excluded from the optional libraries
by listing them in an 'exclude_list' file in the
'config' subdirectory. Consult README.config for
more information.

The initialization task initializes and runs Cexp
as a powerful shell. The idea is that GeSys be used
to load and start 'real' applications at run-time.

RTEMS GeSys has been tested on the following BSPs:

	powerpc-svgm, powerpc-mvme23xx, i386-pc?86, m68k-mvme167

Till Straumann, 2003/7/22

http://www.slac.stanford.edu/~strauman/rtems/index.html


ChangeLog
	2011-08-12 (T.S.):

	* addpath.c: replaced 64-byte path length limit (introduced probably for
	testing) by MAXPATHLEN).

	2011-08-08 (T.S.):

	* configure.ac: modified 'qemuprep' pattern (-> 'qemuprep*')
	so that variants of the BSP are accepted.

	2011-08-04 (T.S.):

	* configure.ac, rtems_netconfig.c: rtems-4.11 has no rtems_bsdnet_loopattach
	anymore. Added 'configure' test and conditional compilation so that we
	can handle 4.11 as well as pre-4.11 versions.

	2011-07-15 (T.S.):

	* configure.ac, config.c: added configure option to enable unified
	work area.

	2011-07-15 (T.S.):

	* cexp-txtregion.c: declare BSP_sbrk_policy as uintptr_t.

	2011-07-15 (T.S.):
	
	* configure.ac: added 'qemuprep' BSP.

	2011-03-17 (T.S.):

	* Makefile.am: add ldep/ to EXTRA_DIST

	2011-02-23 (T.S.):

	* config.c: include <bsp.h> so that BSP-specific idle task (e.g. uC5282)
	is picked up by <confdefs.h>.

	2011-02-03 (T.S.):

	* config.c: added idle task body for qemu selecting sleep mode so
	that host OS doesn't spin (IIRC requires qemu patch since MSR_POW
	is not handled correctly by qemu).

	It's CONFIGURE_FILESYSTEM_NFS (not NFSFS) after all.

	* configure.ac:
	 o select use MEMORY_HUGE for qemu also
	 o set RTEMS_CEXP_TEXT_REGION_SIZE for psim also.
	 o moved libbsdport check away from pcx86 specific section into
	   generic flow of the script.

	* m4/rtems-bsp-postlink.m4: Use PReP bootloader for qemu also.

	* nvram/pairxtract.c: Skip options (stuff starting with '-').

	2010-12-14 (T.S.):

	* configure.ac: default NIC name for mvme2xxx must be 'dc1', not 'dc0'.

	2010-10-05 (T.S.):

	* README.addpath, addpath.c: added comments describing new
	%P substitution.

	2010-10-05 (T.S.):

	* addpath.c: Added %P (cwd) substitution to 'pathSubstitute()'
	(suggestion by K. Luchini).

	BUGFIX: 'stringSubstitute()' failed to add terminal 'NUL' to 
	generated string.

	Dynamically allocate storage for getcwd() buffer.

	2010-10-05 (T.S.):

	* README.addpath: added README.addpath describing addpath & friends.

	2010-06-30 (T.S.):

	* configure.ac: Check for nfsInit rather than nfsMount when looking for
	bundled NFS support (nfsMount is gone from rtems-4.10).

	* Makefile.am: Base BSP specific files and CPPFLAGS on RTEMS_BSP_CLASS
	rather than the RTEMS_BSP.

	* bug_disk.c: Silence compiler warning (rtems-4.10) by including
	<unistd.h>

	* config.c, init.c: definition CONFIGURE_FILESYSTEM_TFTPFS and
	CONFIGURE_FILESYSTEM_NFSFS in config.c replaces explicit initialization
	of TFTPFS and NFS from init.c under rtems-4.10.

	* init.c: reverted declaration of cmdlinetmp as 'const char *' and
	added a comment that we really violate the 'const' attribute here
	for BSPs that declare a 'const' command-line since we do modify it.

	* reboot5282.c: More API mess. From some 4.10 pre-release on
	bsp_reset() is declared in <bsp/bootcard.h>. As of 4.9.99 it is
	bsp_reset(void) [more consistent], before it was bsp_reset(int).

	* psim_tree.gesys: Updated tree by adding OpenPIC and (optional)
	ethernet device (requires patched psim).

	2010-06-24 (T.S.):

	* configure.ac: Added a check for libbsdport (pc?86). If this library is not
	found then with_nic_name/with_nic_attach remain undefined and the BSP
	default NIC will be used (allows to be built w/o libbsdport).

	2010-04-22 (T.S.):

	* configure.ac, Makefile.am: Override clock rate from config.h rather
	than the Makefile.

	2010-04-22 (T.S.):

	* config.c, Makefile.am: Made clock rate 'overridable' from Makefile.am
	and use slower rate (100Hz) for uC5282 BSP.

	2010-04-22 (T.S.):

	* ChangeLog: Added

	* config.c: INCREASED SYSTEM CLOCK RATE TO 1kHz (and reduced timeslice
	to 20ms).


config.c
/*
 * RTEMS configuration for EPICS
 *  rtems_config.c,v 1.1 2001/08/09 17:54:04 norume Exp
 *      Author: W. Eric Norum
 *              eric@cls.usask.ca
 *              (306) 966-6055
 */

#include <rtems.h>
#include <bsp.h>

#ifdef HAVE_CONFIG_H
#include <config.h>
#else
#include "verscheck.h"
#endif

#include <bspopts.h>

/*
 ***********************************************************************
 *                         RTEMS CONFIGURATION                         *
 ***********************************************************************
 */

/*
 * pre 4.10:
#define STACK_CHECKER_ON 1
 * 4.10+:
#define CONFIGURE_STACK_CHECKER_ENABLED
*/

#define CONFIGURE_RTEMS_INIT_TASKS_TABLE

#ifdef MEMORY_SCARCE
#define CONFIGURE_EXECUTIVE_RAM_SIZE        MEMORY_SCARCE
#elif defined MEMORY_HUGE
#define CONFIGURE_EXECUTIVE_RAM_SIZE        (15*1024*1024)
#else
#define CONFIGURE_EXECUTIVE_RAM_SIZE        (5*1024*1024)
#endif
#define CONFIGURE_MAXIMUM_TASKS             rtems_resource_unlimited(30)
#define CONFIGURE_MAXIMUM_SEMAPHORES        rtems_resource_unlimited(500)
#define CONFIGURE_MAXIMUM_TIMERS            rtems_resource_unlimited(20)
#define CONFIGURE_MAXIMUM_MESSAGE_QUEUES    rtems_resource_unlimited(5)
#define CONFIGURE_MAXIMUM_PERIODS		    rtems_resource_unlimited(8)
#define CONFIGURE_MAXIMUM_DRIVERS			15
#define CONFIGURE_MAXIMUM_REGIONS			3
#define CONFIGURE_MAXIMUM_USER_EXTENSIONS	4

#ifdef USE_POSIX
#define CONFIGURE_MAXIMUM_POSIX_THREADS			20
#define CONFIGURE_MAXIMUM_POSIX_MUTEXES			20
#define CONFIGURE_MAXIMUM_POSIX_CONDITION_VARIABLES	20
#define CONFIGURE_MAXIMUM_POSIX_KEYS			20
#define CONFIGURE_MAXIMUM_POSIX_TIMERS			20
#define CONFIGURE_MAXIMUM_POSIX_QUEUED_SIGNALS	20
#define CONFIGURE_MAXIMUM_POSIX_MESSAGE_QUEUES	20
#define CONFIGURE_MAXIMUM_POSIX_SEMAPHORES		20
#endif

#ifdef USE_UNIFIED_WORKAREA
#define CONFIGURE_UNIFIED_WORK_AREAS
#endif

#define CONFIGURE_LIBIO_MAXIMUM_FILE_DESCRIPTORS 512
#define CONFIGURE_USE_IMFS_AS_BASE_FILESYSTEM

#if RTEMS_VERSION_ATLEAST(4,9,99)
#ifdef TFTP_SUPPORT
#define CONFIGURE_FILESYSTEM_TFTPFS
#endif
#ifdef NFS_SUPPORT
#define CONFIGURE_FILESYSTEM_NFS
#endif
#endif

/*****  Allow override from config.h ******/
#ifndef CONFIGURE_MICROSECONDS_PER_TICK
#define CONFIGURE_MICROSECONDS_PER_TICK 1000
#endif

/* Timeslice period = 20ms                */
#define CONFIGURE_TICKS_PER_TIMESLICE   (20000/CONFIGURE_MICROSECONDS_PER_TICK)

#define CONFIGURE_INIT_TASK_PRIORITY    80

#define CONFIGURE_INIT_TASK_INITIAL_MODES (RTEMS_PREEMPT | \
                    RTEMS_NO_TIMESLICE | \
                    RTEMS_NO_ASR | \
                    RTEMS_INTERRUPT_LEVEL(0))
#define CONFIGURE_INIT_TASK_ATTRIBUTES (RTEMS_FLOATING_POINT | RTEMS_LOCAL)
#ifdef MEMORY_SCARCE
#define CONFIGURE_INIT_TASK_STACK_SIZE  (50*1024)
#else
#define CONFIGURE_INIT_TASK_STACK_SIZE  (100*1024)
#endif
rtems_task Init (rtems_task_argument argument);

#define CONFIGURE_HAS_OWN_DEVICE_DRIVER_TABLE

#if RTEMS_VERSION_ATLEAST(4,6,99)
#include <rtems/console.h>
#include <rtems/clockdrv.h>
#ifdef USE_RTC_DRIVER
#include <rtems/rtc.h>
#endif
#else
#include <console.h>
#include <clockdrv.h>
#ifdef USE_RTC_DRIVER
#include <rtc.h>
#endif
#endif

rtems_driver_address_table Device_drivers[]={
    CONSOLE_DRIVER_TABLE_ENTRY,
    CLOCK_DRIVER_TABLE_ENTRY,
#ifdef USE_RTC_DRIVER
	RTC_DRIVER_TABLE_ENTRY,
#endif
    {0}
};


#define CONFIGURE_INIT
#if RTEMS_VERSION_ATLEAST(4,6,99)
#include <rtems/confdefs.h>
#else
#include <confdefs.h>
#endif
